File: camerax_ios/ios/Classes/ImageProxy.swift
//
//  File.swift
//  camerax_ios
//
//  Created by 闫守旺 on 2024/7/13.
//

import Foundation
import AVFoundation

@objc public class ImageProxy: NSObject {
    let imageBuffer: CVImageBuffer
    
    @objc public let format: ImageFormat
    @objc public let width: Int
    @objc public let height: Int
    @objc public let rotationDegrees: Int
    @objc public let planeProxies: [PlaneProxy]
    @objc public let timestamp: Int
    
    init(imageBuffer: CVImageBuffer, format: ImageFormat, width: Int, height: Int, rotationDegrees: Int, planeProxies: [PlaneProxy], timestamp: Int) {
        self.imageBuffer = imageBuffer
        self.format = format
        self.width = width
        self.height = height
        self.rotationDegrees = rotationDegrees
        self.planeProxies = planeProxies
        self.timestamp = timestamp
    }
    
    @objc public func close() {
        CVPixelBufferUnlockBaseAddress(imageBuffer, .readOnly)
    }
}

File: camerax_linux/lib/src/core/image_proxy.dart
import 'dart:ui';

import 'package:camerax_platform_interface/camerax_platform_interface.dart';

class ImageProxyImpl implements ImageProxy {
  @override
  final int timestamp;
  @override
  final ImageFormat format;
  @override
  final int width;
  @override
  final int height;
  @override
  final int rotationDegrees;
  @override
  final List<PlaneProxy> planeProxies;
  final VoidCallback onClosed;

  bool _closed;

  ImageProxyImpl({
    required this.timestamp,
    required this.format,
    required this.width,
    required this.height,
    required this.rotationDegrees,
    required this.planeProxies,
    required this.onClosed,
  }) : _closed = false;

  @override
  void close() {
    final closed = _closed;
    if (closed) {
      return;
    }
    onClosed();
    _closed = true;
  }
}
